File: Data Structures & Algorithms/reorder-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
  public void reorderList(ListNode head) {
    if (head == null || head.next == null) return;

    Stack<ListNode> stack = new Stack<>();
    int length = 0;
    ListNode curr = head;

    while (curr != null) {
      stack.add(curr);
      curr = curr.next;
      length++;
    }

    for (int i = length / 2; i > 0; i--) {
      ListNode next = head.next;
      ListNode popped = stack.pop();

      head.next = popped;
      popped.next = next;

      head = next;
    }
    head.next = null;
  }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        arr = self.to_list(head)
        del arr[-n]
        return self.re_link(arr) if len(arr) > 0 else None

    def to_list(self, head) -> list[ListNode]:
        curr = head
        res = []
        while curr is not None:
            res.append(curr)
            curr = curr.next
        return res

    def re_link(self, arr: list[ListNode]):
        head = arr[0]
        for i, node in enumerate(arr):
            if len(arr) > i + 1:
                node.next = arr[i + 1]
            else:
                node.next = None
        return head
